src-electron/rest/user-data.js

```diff
@@ -697,6 +697,81 @@ function httpDeleteSessionPackage(db) {
   }
 }
 
+/**
+ * Creating a duplicate for endpoint
+ * 
+ * @param {*} db
+ * @returns newly created endpoint id
+ */
+function httpPostDuplicateEndpoint(db) {
+  return async (req, res) => {
+    let endpointId = req.body.id
+    let endpointIdentifier = req.body.endpointIdentifier
+    let endpointTypeId = req.body.endpointTypeId
+    try {
+      let id = await queryEndpoint.duplicateEndpoint(
+        db,
+        endpointId,
+        endpointIdentifier,
+        endpointTypeId
+      )
+      res.status(StatusCodes.OK).json({id:id})
+    } catch (err) {
+      res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(err)
+    }
+  }
+}
+
+/**
+ * Creating a duplicate for endpoint-type and endpoint-type-attributes
+ * 
+ * @param {*} db
+ * @returns newly created endpoint-type id
+ */
+function httpPostDuplicateEndpointType(db) {
+  return async (request, response) => {
+    let { endpointTypeId } = request.body
+    try {
+      let newId = await queryConfig.duplicateEndpointType(
+        db, endpointTypeId
+      )
+
+      duplicateEndpointTypeClusters(db, endpointTypeId, newId)
+
+      response.status(StatusCodes.OK).json({
+        id: newId
+      })
+    } catch (err) {
+      response.status(StatusCodes.INTERNAL_SERVER_ERROR).json(err)
+    }
+  }
+}
+
+/**
+ * duplicate all clusters and attributes of an old endpoint type, using oldEndpointType id and newly created endpointType id
+ * 
+ * @param {*} db
+ * @param {*} oldEndpointTypeId
+ * @param {*} newEndpointTypeId
+ */
+async function duplicateEndpointTypeClusters(db, oldEndpointTypeId, newEndpointTypeId) {
+  try {
+      let oldEndpointTypeClusters = await queryConfig.selectEndpointClusters(db, oldEndpointTypeId);
+      oldEndpointTypeClusters.forEach(async (endpointTypeCluster) => {
+        let newEndpointTypeClusterId = await queryConfig.insertOrReplaceClusterState(db,newEndpointTypeId,endpointTypeCluster.clusterRef,endpointTypeCluster.side, endpointTypeCluster.enabled)
+        let oldAttributes = await queryAttribute.selectEndpointTypeAttributesByEndpointTypeRefAndClusterRef(db,oldEndpointTypeId,endpointTypeCluster.endpointTypeClusterId)
+        oldAttributes.forEach(async (attrubute) => {
+          await queryAttribute.duplicateEndpointTypeAttribute(db,
+            newEndpointTypeId,
+            newEndpointTypeClusterId,
+            attrubute)
+        })
+      })
+  } catch (err) {
```

Comment:
Instead of catching the error here, the exception should just bubble to httpPostDuplicateEndpointType() and the error should be printed  / responded there. otherwise httpPostDuplicateEndpointType() would just fail silently.